File: src/handlers/app_installation_handler.py
import logging
import os
from typing import Dict, Any
from src.handlers.base_handler import CommandHandler
import socketio
from src.services.service_status import ServiceStatus

logger = logging.getLogger(__name__)

class AppInstallationHandler(CommandHandler):
    """Handles app installation requests - clones and prepares build setup without building"""
    
    def __init__(self):
        """
        Initialize the app installation handler.
        """
        # Get service_manager from base class
        service_manager = self.get_service_manager()
        if not service_manager:
            logger.error("No service manager available for AppInstallationHandler")
            self.service_manager = None
            self.build_service = None
            self._socket_manager = None
            self.github_clone_service = None
            return
            
        self.service_manager = service_manager
        self.build_service = service_manager.build_service
        self._socket_manager = service_manager.socket_manager
        self.github_clone_service = service_manager.github_clone_service
        
    def get_command_name(self) -> str:
        return "install_app"
    
    @property
    def socket_manager(self):
        """
        Gets the socket manager, trying to refresh from service_manager if it's None.
        
        Returns:
            The socket manager instance or None if not available
        """
        if self._socket_manager is None:
            # Try to get socket_manager from service_manager
            service_manager = self.get_service_manager()
            if service_manager:
                self._socket_manager = service_manager.socket_manager
        return self._socket_manager
        
    async def get_github_repo_info(self, service_uid: str) -> Dict[str, Any]:
        """
        Gets GitHub repository information for a service from the server.
        
        Args:
            service_uid: ID of the service to get GitHub repo info for
            
        Returns:
            Dictionary containing the GitHub repo info or error information
        """
        socket_manager = self.socket_manager
        if not socket_manager or not hasattr(socket_manager, 'sio'):
            logger.warning(f"No socket manager available to get GitHub repo info for {service_uid}")
            return {'error': "Socket manager not available"}
            
        try:
            result = await socket_manager.sio.call(
                'get_github_repo',
                {'service_uid': service_uid},
                namespace=socket_manager.namespace,
                timeout=30
            )
            
            # Return the result directly - error checking happens in the caller
            return result
            
        except socketio.exceptions.TimeoutError:
            error_msg = f"Timeout getting GitHub repo info for {service_uid}"
            logger.error(error_msg)
            return {'error': error_msg}
        except Exception as e:
            error_msg = f"Failed to get GitHub repo info: {str(e)}"
            logger.error(f"Error getting GitHub repo info for {service_uid}: {e}")
            return {'error': error_msg}
        
    async def handle(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Handle the app installation request - clones and prepares build without building
        
        Args:
            data: Dictionary containing:
                - service_uid: ID of the service to install
                - service: Service data
                
        Returns:
            Dictionary containing the installation result
        """
        if not self.service_manager:
            return {'success': False, 'error': 'Service manager not available'}
            
        try:
            logger.info(f"Installing app with data: {data}")
            service_uid = data.get('service_uid')
            
            if not service_uid:
                raise ValueError("Missing required parameter: service_uid")                
            
            # Check if repo is already cloned
            clone_path = os.path.join(self.github_clone_service.base_clone_dir, service_uid)
            if os.path.exists(clone_path):
                logger.info(f"Repository for service {service_uid} already exists at {clone_path}")
                # Refresh the build setup - not an initial clone
                await self.build_service.prepare_build(clone_path, service_uid, is_initial_clone=False)
                return {'success': True, 'clone_path': clone_path}
            
            # Repository doesn't exist, get repo info and clone
            try:
                repo_info = await self.get_github_repo_info(service_uid)
                
                if not repo_info or not isinstance(repo_info, dict) or 'error' in repo_info:
                    error_msg = repo_info.get('error') if isinstance(repo_info, dict) else "Invalid response"
                    logger.error(f"Error getting GitHub repo info for {service_uid}: {error_msg}")
                    await self.service_manager.status_service.update_service_status(service_uid, ServiceStatus.FAILED, f"Failed to get repo info: {error_msg}")
                    return {'success': False, 'error': error_msg}

                repo_url = repo_info.get('repo_url')
                access_token = repo_info.get('access_token')
                if not repo_url or not access_token:
                    error_msg = f"Missing repo_url or access_token in response for {service_uid}"
                    logger.error(f"{error_msg}: {repo_info}")
                    await self.service_manager.status_service.update_service_status(service_uid, ServiceStatus.FAILED, error_msg)
                    return {'success': False, 'error': error_msg}
                    
                # Clone Repo
                clone_result = await self.github_clone_service.clone_repo(repo_url, access_token, service_uid)
                
                if not clone_result or not clone_result.get("success"):
                    error_msg = clone_result.get('error', 'Unknown clone error')
                    logger.error(f"Clone failed for {service_uid}: {error_msg}")
                    await self.service_manager.status_service.update_service_status(service_uid, ServiceStatus.FAILED, error_msg)
                    return {'success': False, 'error': error_msg}
                
                retrieved_clone_path = clone_result.get('clone_path')
                
                # The GitHub clone service typically calls prepare_build automatically, 
                # but we'll ensure it happens for consistent behavior
                # Since this is a fresh clone, set is_initial_clone=True
                if retrieved_clone_path:
                    await self.build_service.prepare_build(retrieved_clone_path, service_uid, is_initial_clone=True)
                
                return {
                    'success': True,
                    'clone_path': retrieved_clone_path
                }
                    
            except Exception as e:
                error_msg = f"Error installing app for {service_uid}: {str(e)}"
                logger.error(error_msg)
                logger.exception(e)
                await self.service_manager.status_service.update_service_status(service_uid, ServiceStatus.FAILED, error_msg)
                return {'success': False, 'error': error_msg}
                
        except Exception as e:
            logger.error(f"Error in app installation: {str(e)}")
            logger.exception(e)
            if 'service_uid' in locals():
                await self.service_manager.status_service.update_service_status(service_uid, ServiceStatus.FAILED, str(e))
            return {'success': False, 'error': str(e)} 

File: internal/handlers/app_installation_handler.go
package handlers

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"

	"pulseup-agent-go/pkg/logger"
	"pulseup-agent-go/pkg/types"

	"github.com/go-git/go-git/v5"
	"github.com/go-git/go-git/v5/plumbing/transport/http"
)

// AppInstallationHandler handles application installation requests
type AppInstallationHandler struct {
	*BaseHandler
}

// AppInstallationRequest represents the request structure for app installation
type AppInstallationRequest struct {
	ServiceUID  string                 `json:"service_uid"`
	AccessToken string                 `json:"access_token"`
	GitHubRepo  string                 `json:"github_repo"`
	Service     map[string]interface{} `json:"service,omitempty"`
}

// NewAppInstallationHandler creates a new app installation handler
func NewAppInstallationHandler(logger *logger.Logger, services ServiceProvider) *AppInstallationHandler {
	return &AppInstallationHandler{
		BaseHandler: NewBaseHandler(logger.With("handler", "app_installation"), services),
	}
}

// Base returns the underlying BaseHandler for routing helpers.
func (h *AppInstallationHandler) Base() *BaseHandler {
	return h.BaseHandler
}

// Install processes the app installation command - clones and prepares build without building
func (h *AppInstallationHandler) Install(ctx context.Context, data json.RawMessage) (*types.CommandResponse, error) {
	var request AppInstallationRequest
	if err := json.Unmarshal(data, &request); err != nil {
		return &types.CommandResponse{
			Success: false,
			Error:   "invalid request format: " + err.Error(),
		}, nil
	}

	// Validate required fields
	if request.ServiceUID == "" {
		return &types.CommandResponse{
			Success: false,
			Error:   "service_uid is required",
		}, nil
	}

	if request.AccessToken == "" {
		return &types.CommandResponse{
			Success: false,
			Error:   "access_token is required",
		}, nil
	}

	if request.GitHubRepo == "" {
		return &types.CommandResponse{
			Success: false,
			Error:   "github_repo is required",
		}, nil
	}

	h.logger.Info("Installing app", "service_uid", request.ServiceUID, "github_repo", request.GitHubRepo)

	// Check if repo is already cloned
	baseCloneDir := "/opt/pulseup/apps"
	if os.Getenv("DEBUG") == "true" {
		if debugDir := os.Getenv("DEBUG_CLONE_DIR"); debugDir != "" {
			baseCloneDir = debugDir
		}
	}
	clonePath := filepath.Join(baseCloneDir, request.ServiceUID)
	if _, err := os.Stat(clonePath); err == nil {
		h.logger.Info("Repository already exists", "service_uid", request.ServiceUID, "clone_path", clonePath)

		// Refresh the build setup - not an initial clone
		if err := h.services.GetBuildService().PrepareBuild(ctx, clonePath, request.ServiceUID, false); err != nil {
			h.logger.Error("Failed to prepare build", "error", err)
			return &types.CommandResponse{
				Success: false,
				Error:   "Failed to prepare build: " + err.Error(),
			}, nil
		}

		return &types.CommandResponse{
			Success: true,
			Data: map[string]interface{}{
				"success":    true,
				"clone_path": clonePath,
			},
		}, nil
	}

	// Repository doesn't exist, clone it directly using provided access token and repo
	cloneResult, err := h.cloneGitHubRepoWithToken(ctx, request.GitHubRepo, request.AccessToken, request.ServiceUID)
	if err != nil {
		h.logger.Error("Failed to clone GitHub repository", "service_uid", request.ServiceUID, "github_repo", request.GitHubRepo, "error", err)
		h.updateServiceStatus(request.ServiceUID, types.ServiceStatusFailed, "Failed to clone repository: "+err.Error())
		return &types.CommandResponse{
			Success: false,
			Error:   err.Error(),
		}, nil
	}

	if !cloneResult.Success {
		h.logger.Error("Clone failed", "service_uid", request.ServiceUID, "github_repo", request.GitHubRepo, "error", cloneResult.Error)
		h.updateServiceStatus(request.ServiceUID, types.ServiceStatusFailed, cloneResult.Error)
		return &types.CommandResponse{
			Success: false,
			Error:   cloneResult.Error,
		}, nil
	}

	h.logger.Info("Repository cloned successfully", "clone_path", cloneResult.ClonePath)

	// Prepare build setup - this is a fresh clone, so set is_initial_clone=true
	if err := h.services.GetBuildService().PrepareBuild(ctx, cloneResult.ClonePath, request.ServiceUID, true); err != nil {
		h.logger.Error("Failed to prepare build", "error", err)
		return &types.CommandResponse{
			Success: false,
			Error:   "Failed to prepare build: " + err.Error(),
		}, nil
	}

	return &types.CommandResponse{
		Success: true,
		Data: map[string]interface{}{
			"success":    true,
			"clone_path": cloneResult.ClonePath,
		},
	}, nil
}

// cloneGitHubRepoWithToken clones a GitHub repository using the provided access token
func (h *AppInstallationHandler) cloneGitHubRepoWithToken(ctx context.Context, githubRepo, accessToken, serviceUID string) (*types.CloneResult, error) {
	h.logger.Info("Cloning GitHub repository with token", "github_repo", githubRepo, "service_uid", serviceUID)

	// Construct the clone path using the service UID
	baseCloneDir := "/opt/pulseup/apps"
	if os.Getenv("DEBUG") == "true" {
		if debugDir := os.Getenv("DEBUG_CLONE_DIR"); debugDir != "" {
			baseCloneDir = debugDir
		}
	}
	clonePath := filepath.Join(baseCloneDir, serviceUID)

	// Check if directory already exists
	if _, err := os.Stat(clonePath); err == nil {
		h.logger.Info("Repository already exists", "clone_path", clonePath)
		return &types.CloneResult{
			Success:   true,
			ClonePath: clonePath,
		}, nil
	}

	// Construct the full GitHub URL from repo name
	repoURL := "https://github.com/" + githubRepo + ".git"

	// Clone the repository with token authentication
	if err := h.cloneWithToken(ctx, repoURL, accessToken, clonePath); err != nil {
		return &types.CloneResult{
			Success: false,
			Error:   "Failed to clone repository: " + err.Error(),
		}, nil
	}

	h.logger.Info("Repository cloned successfully", "clone_path", clonePath)
	return &types.CloneResult{
		Success:   true,
		ClonePath: clonePath,
	}, nil
}

// cloneWithToken clones a repository using token authentication
func (h *AppInstallationHandler) cloneWithToken(ctx context.Context, repoURL, accessToken, clonePath string) error {
	h.logger.Debug("Cloning repository with token", "repo_url", repoURL, "clone_path", clonePath)

	// Ensure the target directory exists
	if err := os.MkdirAll(clonePath, 0755); err != nil {
		return fmt.Errorf("failed to create clone directory: %w", err)
	}

	// Clone the repository with token authentication
	cloneOptions := &git.CloneOptions{
		URL: repoURL,
		Auth: &http.BasicAuth{
			Username: "oauth2",
			Password: accessToken,
		},
	}

	_, err := git.PlainCloneContext(ctx, clonePath, false, cloneOptions)
	if err != nil {
		return fmt.Errorf("git clone failed: %w", err)
	}

	return nil
}

// updateServiceStatus updates the service status using the status service
func (h *AppInstallationHandler) updateServiceStatus(serviceUID string, status types.ServiceStatus, errorMessage string) {
	statusService := h.services.GetStatusService()
	if statusService == nil {
		h.logger.Warn("Status service not available for status update", "service_uid", serviceUID)
		return
	}

	ctx := context.Background()
	if err := statusService.UpdateServiceStatus(ctx, serviceUID, status, errorMessage); err != nil {
		h.logger.Error("Failed to update service status", "service_uid", serviceUID, "status", status, "error", err)
	}
}

// updateStatus sends status updates via WebSocket (legacy method)
func (h *AppInstallationHandler) updateStatus(serviceUID, status, message string) {
	wsManager := h.services.GetWebSocketManager()
	if wsManager == nil {
		h.logger.Warn("WebSocket manager not available for status update")
		return
	}

	statusUpdate := map[string]interface{}{
		"service_uid": serviceUID,
		"status":      status,
		"message":     message,
		"timestamp":   h.getCurrentTimestamp(),
	}

	if err := wsManager.Emit("service_status_update", statusUpdate); err != nil {
		h.logger.Error("Failed to send status update", "error", err)
	}
}

// getCurrentTimestamp returns the current Unix timestamp
func (h *AppInstallationHandler) getCurrentTimestamp() int64 {
	return time.Now().Unix()
}
